api/jobs/handlers.py

```diff
@@ -271,3 +256,128 @@ def retry(self, _id):
+            # Don't enforce permissions for superuser requests or drone requests
+            query = {}
+        else:
+            query = {'origin.id': self.uid}
+        return batch.get_all(query, {'proposed_inputs':0})
+
+    @require_login
+    def get(self, _id):
+        """
+        Get a batch job by id.
+        Use param jobs=true to replace job id list with job objects.
+        """
+
+        get_jobs = self.is_true('jobs')
+        batch_job = batch.get(_id, projection={'proposed_inputs':0}, get_jobs=get_jobs)
+        self._check_permission(batch_job)
+        return batch_job
+
+    @require_login
+    def post(self):
+        """
+        Create a batch job proposal, insert as 'pending'.
+        """
+
+        if self.superuser_request:
+            # Don't enforce permissions for superuser requests
+            user = None
+        else:
+            user = {'_id': self.uid, 'site': self.user_site}
+
+        payload = self.request.json
+        gear_name = payload.get('gear')
+        targets = payload.get('targets')
+        config_ = payload.get('config', {})
+
+        if not gear_name or not targets:
+            self.abort(400, 'A gear name and list of target containers is required.')
+        gear = get_gear_by_name(gear_name)
+        validate_gear_config(gear, config_)
+
+        container_ids = []
+        container_type = None
+
+        for t in targets:
+            if not container_type:
+                container_type = t.get('type')
+            else:
+                # Ensure all targets are of same type, may change in future
+                if container_type != t.get('type'):
+                    self.abort(400, 'targets must all be of same type.')
+            container_ids.append(t.get('id'))
+
+        objectIds = [bson.ObjectId(x) for x in container_ids]
+        containers = ContainerStorage.factory(container_type).get_all_el({'_id': {'$in':objectIds}}, user, {'permissions': 0})
+
+        if len(containers) != len(container_ids):
+            # TODO: Break this out into individual errors, requires more work to determine difference
+            self.abort(404, 'Not all target containers exist or user does not have access to all containers.')
+
+
+        results = batch.find_matching_conts(gear, containers, container_type)
+
+        matched = results['matched']
+        if not matched:
+            self.abort(400, 'No target containers matched gear input spec.')
+
+        batch_proposal = {
+            '_id': bson.ObjectId(),
+            'gear': gear_name,
+            'config': config_,
+            'state': 'pending',
+            'origin': self.origin,
+            'proposed_inputs': [c.pop('inputs') for c in matched]
+        }
+
+        batch.insert(batch_proposal)
+        batch_proposal.pop('proposed_inputs')
+        batch_proposal['not_matched'] = results['not_matched'],
+        batch_proposal['ambiguous'] = results['ambiguous'],
+        batch_proposal['matched'] = matched
```

Comment:
I remember when it's separate - these are not stored on the batch object in mongo, only returned so the caller can display/do something intelligent with the containers that match/didn't match/were ambiguous. The batch job doesn't need to know this information.